src/main/java/software/amazon/awssdk/crt/http/HttpStreamManager.java

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.awssdk.crt.http;
+
+import software.amazon.awssdk.crt.CrtRuntimeException;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Manages a pool for either HTTP/1.1 or HTTP/2 connection.
+ * For HTTP/1.1, it will grab a connection from HttpClientConnectionManager to
+ * make request on it, and will return it back until the request finishes.
+ * For HTTP/2, it will use Http2StreamManager under the hood.
+ */
+public class HttpStreamManager implements AutoCloseable {
+
+    private HttpClientConnectionManager connectionManager = null;
+    private Http2StreamManager h2StreamManager = null;
+    private CompletableFuture<Void> shutdownComplete = null;
+    private AtomicLong shutdownNum = new AtomicLong(0);
+    private Throwable shutdownCompleteException = null;
+
+    /**
+     * Factory function for HttpStreamManager instances
+     *
+     * @param options configuration options
+     * @return a new instance of an HttpStreamManager
+     */
+    public static HttpStreamManager create(HttpStreamManagerOptions options) {
+        return new HttpStreamManager(options);
+    }
+
+    private HttpClientConnectionManagerOptions getConnManagerOptFromStreamManagerOpt(HttpStreamManagerOptions options) {
+
+        HttpClientConnectionManagerOptions connManagerOptions = new HttpClientConnectionManagerOptions();
+        connManagerOptions.withClientBootstrap(options.getClientBootstrap())
+                .withSocketOptions(options.getSocketOptions())
+                .withTlsContext(options.getTlsContext())
+                .withWindowSize(options.getWindowSize())
+                .withUri(options.getUri())
+                .withMaxConnections(options.getMaxConnections())
```

Comment:
instead of max connections, maybe something like JAVA SDK, using max concurrency that will different between protocols. https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java#L267-L272